### ISSUE TITLE
Awaitablity checks not fully complete

### ISSUE THREAD

**Issue by Gobot1234** [microsoft/pyright#6053]
**Describe the bug**
Pyright's checks for whether an object is awaitable are not fully in line with python's. This code runs fine at runtime but type errors

**Code or Screenshots**
```python
import asyncio
import collections.abc
import contextvars
from typing import Self


class WrapCoro(collections.abc.Coroutine):
    def __init__(self, coro, context):
        self._coro = coro
        self._context = context

    def __await__(self) -> Self:
        return self

    def __iter__(self) -> Self:
        return self

    def __next__(self):
        return self.send(None)

    def throw(self, *exc_info):
        result = self._context.run(self._coro.throw, *exc_info)
        return result

    def send(self, v):
        result = self._context.run(self._coro.send, v)
        return result

async def coro():
    ...

async def foo():
    await WrapCoro(coro(), contextvars.copy_context())  # warning that WrapCoro is not awaitable

asyncio.run(foo())
```

**VS Code extension or command-line**
pyright 1.1.328

**Comment by erictraut** [comment]
Your class derives from an abstract base class, but your implementation isn't compatible with the ABC. The `collections.abc.Coroutine` requires you to supply an `__await__` method that returns a `Generator[Any, None, _T_co]`, but your implementation returns `Self`. Was your intent for `WrapCoro` to also derive from `Generator`?

**Comment by Gobot1234** [comment]
Even if it does derive from Generator, pyright still reports an error. Should inheritance even be necessary, should these things not be checked based on Protocol compatibility?

**Comment by erictraut** [comment]
Which protocol do you suggest?

**Comment by Gobot1234** [comment]
I'll open a typeshed issue about it. It seems like the collections.abc protocols aren't actually in line with what you actually get at runtime anyway and you should be using the types classes if you want some of the fancier attributes

**Comment by erictraut** [comment]
I changed pyright's logic to use the `typing.Awaitable` protocol. It's not ideal because it relies on the `Generator` class, which (as you've pointed out in other threads) imposes requirements that the runtime does not, but it's an improvement over pyright's (and mypy's) current logic. It addresses this issue.

If the `Awaitable` protocol is updated in typeshed in the future, pyright will automatically take advantage of the improved definition.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.330, which I just published. It will also be included in an upcoming release of pylance.
